### PR TITLE
lego: 4.35.2 -> 5.4.1

### LINKED ISSUE #529238: Update Request: lego 4.35.2 → 5.2.2
### Nixpkgs version

- Unstable (26.11)

### Package name

lego

### Upstream version

5.2.2

### Nixpkgs version

4.35.2

### Changelog

https://github.com/go-acme/lego/blob/main/CHANGELOG.md

### Additional context

This is the currently used ACME client for Let's Encrypt certificates. Good real-world testing may be required before doing major upgrades like this.

### Notify maintainers

@aanderse @arianvp @emilazy @m1cr0man
---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)

### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this package update does not yet exist in an [open pull request](https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+is%3Apr+label%3A%228.has%3A+package+%28update%29%22) or in [Nixpkgs Unstable](https://search.nixos.org/packages?channel=unstable).
- [x] I assert that this is not a [duplicate of any known issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+label%3A%229.needs%3A+package+%28update%29%22).
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.
- [x] I assert that I have read the [automation/AI policy](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#automationai-policy) and that this issue report complies with it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by geri1701**
Update `lego` to 5.4.1 and migrate the NixOS ACME integration to lego 5. This updates the generated `lego run` invocations, migrates existing account state, and documents the raw `extraLego*Flags` CLI compatibility change in the release notes.

See the [upstream changelog](https://github.com/go-acme/lego/blob/v5.4.1/CHANGELOG.md) for release details.

The existing ACME VM test matrix passes, with added coverage for handling v4 account and certificate state.

Closes #529238. Follow-up to #529467.

Assisted-by: pi coding agent / Mika (OpenAI gpt-5.6-sol)

## Things done

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [x] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [x] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [x] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [x] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.
- [x] Follows the [automation/AI policy].

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[automation/AI policy]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#automationai-policy
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

### CODE REVIEW

**Comment by geri1701** [comment]
@NixOS/acme gentle ping when you have time. This updates `lego` to 5.2.2 and adapts `security.acme`; GitHub Actions and `no PR failures` are green.

File: nixos/modules/security/acme/default.nix [hunk 1]
```diff
@@ -254,7 +254,10 @@ let
               "--dns"
               data.dnsProvider
             ]
-            ++ lib.optionals (!data.dnsPropagationCheck) [ "--dns.propagation-disable-ans" ]
+            ++ lib.optionals (!data.dnsPropagationCheck) [
+              "--dns.propagation.wait"
```
Comment (debtquity) [inline nixos/modules/security/acme/default.nix]: Per lego 5 migration guide [1], `--dns.propagation-disable-ans` was renamed to `--dns.propagation.disable-ans`. Seems better to keep the original behavior, rather than use a different option.

On further Inspection of `--dns.propagation.wait` [2], `0s` is the default value:

> By setting this flag, disables all the propagation checks of the TXT record and uses a wait duration instead. (Default: 0s)

So, in current state of PR. `lego` will still perform active authoritative and recursive checks on the TXT record even if the user specifies `dnsPropagationCheck = false`.

This is confirmed by inspection of upstream code [3].

If you want leverage this option, have to use a value of `1s` or greater to disable the active authoritative and recursive checks to rely on an interval check. And you would have to change `dnsPropagationCheck` option so that it's more ergonomic to the end user (ie, avoid hardcoding `1s` into this module).

* [1] https://go-acme.github.io/lego/migration/cli/index.html
* [2] https://go-acme.github.io/lego/references/ref-flags/index.html#flags-related-to-the-dns-01-challenge
* [3] https://github.com/go-acme/lego/blob/v5.4.1/cmd/internal/root/process_challenges.go#L154-L156

File: nixos/modules/security/acme/default.nix [hunk 2]
```diff
@@ -309,21 +312,24 @@ let
...
-      # Although --must-staple is common to both modes, it is not declared as a
-      # mode-agnostic argument in lego and thus must come after the mode.
       runOpts = lib.escapeShellArgs (
-        commonOpts
-        ++ [ "run" ]
+        [ "run" ]
+        ++ commonOpts
+        ++ [
+          "--no-random-sleep"
+          "--renew-force"
```
Comment (debtquity) [inline nixos/modules/security/acme/default.nix:325]: If `runOpts` are only invoked on initial certificate creation, why is `--ari-disable` necessary here? Especially in the presence of `--renew-force`.

**Comment by greg-hellings** [comment]
Giving this a run through because 5.0 introduced the `--ipv4only` option which hopefully addresses a constant timeout I'm seeing when the client tries to query DNS over IPv6

**Review by greg-hellings — APPROVED** [review]
Tested working and providing me needed function.

Upgrade was clean and without issue when I moved from HTTP-01 challenge to DNS-01 challenge, which necessitated moving to this for the new features in version 5.0 for my use case.

**Comment by geri1701** [comment]
> Tested working and providing me needed function.

Thanks for testing and reporting back!

**Comment by aanderse** [comment]
would be really nice to get this merged... but i think we really need @m1cr0man to approve this before we can have the confidence we would like...

any chance you can give it a look over @m1cr0man? :bowing_man:

**Review by greg-hellings — APPROVED** [review]

**Comment by geri1701** [comment]
I’m very confident overall. If there is a specific concern, I can extend the test coverage.

**Review by greg-hellings — DISMISSED** [review]

**Comment by greg-hellings** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 539076`
Commit: `3139752974b0e2238965b319b793b21f08c71caf`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>lego</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>

**Review by bachp — DISMISSED** [review]
I deployed this and it works. It also fixes an issue with the infomaniak.ch DNS provider v2.

**Comment by geri1701** [comment]
> I deployed this and it works. It also fixes an issue with the infomaniak.ch DNS provider v2.

Thank you for testing and confirming that it works!

**Review by felbinger — DISMISSED** [review]

**Review by SuperSandro2000 — DISMISSED** [review]

File: nixos/modules/security/acme/default.nix [hunk 3]
```diff
@@ -254,7 +254,10 @@ let
               "--dns"
               data.dnsProvider
             ]
-            ++ lib.optionals (!data.dnsPropagationCheck) [ "--dns.propagation-disable-ans" ]
+            ++ lib.optionals (!data.dnsPropagationCheck) [
+              "--dns.propagation.wait"
+              "0s"
+            ]
```
Comment (SuperSandro2000) [inline nixos/modules/security/acme/default.nix:262]:
```suggestion
            ++ lib.optionals (!data.dnsPropagationCheck) [
              "--dns.propagation.disable-ans"
            ]
```

File: nixos/modules/security/acme/default.nix [hunk 4]
```diff
@@ -575,6 +581,12 @@ let
...
 
+          # lego v5 moved account keys out of the keys/ subdirectory.
+          # Guard the migration because lego migrate is not idempotent.
+          if [ -n "$(find accounts -path '*/keys/*.key' -print -quit)" ]; then
+            printf 'Y\n' | lego migrate --account-only --path .
+          fi
+
           echo '${domainHash}' > domainhash.txt
 
           # Check if a new order is needed
```
Comment (SuperSandro2000) [inline nixos/modules/security/acme/default.nix]: Does the --force-cert-domains flag replace the condition 4 lines below this comment where we try to check if the domain is still the same?

File: nixos/modules/security/acme/default.nix [hunk 5]
```diff
@@ -309,21 +312,24 @@ let
...
+          "--no-random-sleep"
+          "--renew-force"
+          "--ari-disable"
+        ]
+        ++ lib.optionals (data.csr == null) [ "--force-cert-domains" ]
         ++ lib.optionals data.ocspMustStaple [ "--must-staple" ]
         ++ lib.optionals (data.profile != null) [ "--profile=${data.profile}" ]
         ++ data.extraLegoRunFlags
       );
       renewOpts = lib.escapeShellArgs (
```
Comment (SuperSandro2000) [inline nixos/modules/security/acme/default.nix:333]: Could we consolidate renewOpts into runOpts or would that break certain edge usecases?

**Review by mweinelt — DISMISSED** [review]
Looking at the changes I am biased against accepting these due to an obvious lack of rigor which can probably be attributed to LLM use.

File: nixos/modules/security/acme/default.nix (hunk 3 above)
Comment (geri1701) [reply nixos/modules/security/acme/default.nix:262]: `dnsPropagationCheck = false` disables lego's propagation check as a whole. 
In lego 5, `--dns.propagation.wait 0s` does that, while `--dns.propagation.disable-ans` only disables the authoritative nameserver check and leaves recursive propagation checking enabled.

File: nixos/modules/security/acme/default.nix (hunk 4 above)
Comment (geri1701) [reply nixos/modules/security/acme/default.nix]: Yes, thanks for catching that. `--force-cert-domains` makes lego renew using exactly the configured domains, so I removed the `domainHash` workaround. I also added a regression test that removes an existing SAN while preserving the remaining IP SAN.

File: nixos/modules/security/acme/default.nix (hunk 5 above)
Comment (geri1701) [reply nixos/modules/security/acme/default.nix:333]: Not without changing behavior. 
`runOpts` is used by the full-run path and intentionally adds `--renew-force`, `--ari-disable`, and `extraLegoRunFlags`. `renewOpts` is used for normal renewal and uses `extraLegoRenewFlags`. Sharing one option set would either force renewal on every timer run or remove flags needed by the full-run path.

**Comment by geri1701** [comment]
> Looking at the changes I am biased against accepting these due to an obvious lack of rigor which can probably be attributed to LLM use.

I disagree with this assessment. Could you clarify what specific changes you are requesting on the current head?

**Review by bachp — APPROVED** [review]

File: nixos/modules/security/acme/default.nix (hunk 3 above)
Comment (greg-hellings) [reply nixos/modules/security/acme/default.nix:262]: Indeed. This change currently is working correctly as is.

**Review by greg-hellings — APPROVED** [review]

**Comment by greg-hellings** [comment]
> Looking at the changes I am biased against accepting these due to an obvious lack of rigor which can probably be attributed to LLM use.

I'm curious as to what you object to, specifically? This is a precise and targeted update, it is working great on my systems in addition to the submitters. Was this objection submitted to the wrong PR?

**Review by debtquity — COMMENTED** [review]
summary:

* questionable use of `--dns.propagation.wait`
* on initial cert request in `runOpts`, why is `--disable-ari` used?

File: nixos/modules/security/acme/default.nix (hunk 1 above)
Comment (geri1701) [reply nixos/modules/security/acme/default.nix]: Good point about keeping the v4 behavior. The linked code applies when lego uses its configuration file. NixOS uses the CLI path. There, `cmd.IsSet(...)` means that an explicit `0s` selects `PropagationWait(0, true)` and skips the active precheck.

The explicit flags match the v4 behavior more closely. In v4, NixOS used `--dns.propagation-disable-ans`, which disabled the authoritative propagation requirement. Recursive propagation was not required by default, and the initial recursive lookup still ran. Because v5 enables both propagation requirements by default, I replaced the wait flag with both disable flags. I also added a test that checks both generated `lego run` commands for these exact flags and verifies that the wait flag is absent.

File: nixos/modules/security/acme/default.nix (hunk 2 above)
Comment (geri1701) [reply nixos/modules/security/acme/default.nix:325]: `runOpts` is not limited to initial issuance. NixOS also uses it when the stored certificate key, certificate, or account key is missing. If lego's certificate resource still exists, `lego run` selects its renewal path.

In that path, when the profile and domains are unchanged, `getARIInfo()` runs before the renewal-period guard that checks `--renew-force`. `--renew-force` only bypasses that later guard. It does not suppress the ARI `replaces` identifier. `--ari-disable` does. A stale identifier can cause the CA to reject the order, as in [#448921](https://redirect.github.com/NixOS/nixpkgs/issues/448921).

With no certificate resource, lego takes the obtain path and neither flag has an effect. The normal `renewOpts` path remains ARI-enabled by default.

File: nixos/tests/acme/http01-builtin.nix [hunk 6]
```diff
@@ -298,6 +375,16 @@ in
           check_permissions(builtin, cert, "acme")
           check_permissions(builtin, cert2, "acme")
 
+      with subtest("Remove an existing cert domain"):
+          builtin.succeed("systemctl stop renew-triggered.target")
```
Comment (m1cr0man) [inline nixos/tests/acme/http01-builtin.nix:497]: Ok this is great - so domainhash.txt is not required anymore.

**Comment by m1cr0man** [comment]
I found a little problem testing this locally. I have a setup involving three certificates on one host. During the migration, one of the three services failed because the account key was already moved. Would you be able to add some sort of lockfile so that the migration is only handled by one running/starting acme service?

**Comment by geri1701** [comment]
Thanks, I reproduced this with three certificates sharing the same account. All three services could pass the legacy-key check before the first `lego migrate` moved the account key.

I added an account-specific `flock` around both the check and the migration. The VM test now starts three certificate services concurrently from v4 account state, asserts that the migration runs exactly once, and verifies that all three certificates remain valid.

File: nixos/tests/acme/http01-builtin.nix (hunk 6 above)
Comment (geri1701) [reply nixos/tests/acme/http01-builtin.nix:497]: Yes. While extending this test, I found three edge cases where lego's exact-domain comparison was not stable: a secondary IDN, a duplicate configured identifier, and a noncanonical IPv6 spelling. The updated code handles all three cases and verifies that a second run without changes reaches lego's `Skip renewal:` path. `domainhash.txt` remains unnecessary.

File: pkgs/by-name/le/lego/package.nix
Comment (m1cr0man) [inline pkgs/by-name/le/lego/package.nix]: Introducing this patch now is a bit interesting. The problem with patches is that now they need to be reviewed on every package update, which can be quite tedious for packages which update frequently (and lego certainly does). Do you plan to upstream this? If so, can you link the PR here when you do?

Comment (m1cr0man) [reply pkgs/by-name/le/lego/package.nix]: Infact, I would go so far as to say we don't care. If someone puts canonical and non canonical IPV6 addresses in their domains, that's on them. It's a lot of burden on us as maintainers for a very small minority of people, and something we can put in documentation until this patch is made upstream.

**Review by m1cr0man — COMMENTED** [review]
The IDN patchwork introduced in the last push is a bit overreaching in my opinion. This is a lot of extra code to avoid an issue which can be resolved by correct configuration. I'm not fond of us patching lego for an issue which would be very uncommon in practice, and avoidable with documentation.

Otherwise the account locking looks good, and all the tests passed locally.

File: pkgs/by-name/le/lego/package.nix
Comment (geri1701) [reply pkgs/by-name/le/lego/package.nix]: Agreed. I removed the IDN backport and the downstream IPv6 patch rather than carrying
 them in nixpkgs. I do not plan to upstream the IPv6 normalization as part of this change.
 I also removed the identifier deduplication and related normalization tests. The NixOS
 26.11 migration note now tells users to use unique, canonical identifiers with IDNA
 A-labels and RFC 5952 notation for IPv6 addresses. The account migration lock and its
 regression test remain. All six ACME VM tests pass locally.

**Review by m1cr0man — APPROVED** [review]
Great, I'm happy with this now :)

**Comment by geri1701** [comment]
> Great, I'm happy with this now :)

Thanks. I'll keep the PR up to date while it is under review.

**Comment by m1cr0man** [comment]
Merging now. I'm confident in the test suite and the implementation. We are 100% going to see people open issues for acme-related things (at the very least, for the option changes), but we just have to be ready for them. Not our first time doing something like this on security.acme - see #77578

**Review by dotlambda — COMMENTED** [review]
Why don't we support the [configuration file](https://go-acme.github.io/lego/references/ref-file/index.html)?

**Comment by geri1701** [comment]
> Why don't we support the [configuration file](https://go-acme.github.io/lego/references/ref-file/index.html)?

The need for configuration-file support was not raised before this PR was merged. For this update, I think it made sense to preserve the existing NixOS module behavior while moving from lego v4 to v5. Supporting the new configuration-file mode would be separate functionality. If there is interest in adding it, I think the module design would be better discussed in an issue before implementation.

**Comment by m1cr0man** [comment]
Agreed. The config file may significantly reduce the acme module complexity, but it would be an equally significant change to the module's options and existing configuration of certs, probably requiring more migration logic. Definitely should be discussed in a new issue
